File: mobile_app/lib/models/order.dart
import 'order_detail.dart';

class Order {
  final int orderId;
  final String invoiceCode;
  final double totalPrice;
  final double subtotalAmount;
  final double shippingCost;
  final double discountAmount;
  final String orderStatus;
  final String orderTime;
  final String? estimatedArrival;
  final String receiverName;
  final String receiverPhone;
  final String shippingAddress;
  final List<OrderDetail> orderDetails;

  Order({
    required this.orderId,
    required this.invoiceCode,
    required this.totalPrice,
    required this.subtotalAmount,
    required this.shippingCost,
    required this.discountAmount,
    required this.orderStatus,
    required this.orderTime,
    this.estimatedArrival,
    required this.receiverName,
    required this.receiverPhone,
    required this.shippingAddress,
    this.orderDetails = const [],
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    var detailsList = json['orderDetails'] as List? ?? [];
    List<OrderDetail> details = detailsList.map((i) => OrderDetail.fromJson(i)).toList();

    return Order(
      orderId: json['orderId'] ?? 0,
      invoiceCode: json['invoiceCode'] ?? '',
      totalPrice: (json['totalPrice'] ?? 0).toDouble(),
      subtotalAmount: (json['subtotalAmount'] ?? 0).toDouble(),
      shippingCost: (json['shippingCost'] ?? 0).toDouble(),
      discountAmount: (json['discountAmount'] ?? 0).toDouble(),
      orderStatus: json['orderStatus'] ?? 'pending',
      orderTime: json['orderTime'] ?? '',
      estimatedArrival: json['estimatedArrival'],
      receiverName: json['receiverName'] ?? '',
      receiverPhone: json['receiverPhone'] ?? '',
      shippingAddress: json['shippingAddress'] ?? '',
      orderDetails: details,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'orderId': orderId,
      'invoiceCode': invoiceCode,
      'totalPrice': totalPrice,
      'subtotalAmount': subtotalAmount,
      'shippingCost': shippingCost,
      'discountAmount': discountAmount,
      'orderStatus': orderStatus,
      'orderTime': orderTime,
      'estimatedArrival': estimatedArrival,
      'receiverName': receiverName,
      'receiverPhone': receiverPhone,
      'shippingAddress': shippingAddress,
      'orderDetails': orderDetails.map((v) => v.toJson()).toList(),
    };
  }
}

File: backend-api/src/main/java/com/ogani/api/model/Order.java
package com.ogani.api.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "orders")
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "order_id")
    private Integer orderId;

    @Column(name = "invoice_code", unique = true, length = 50)
    private String invoiceCode;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "address_id")
    private UserAddress address;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "payment_method_id")
    private PaymentMethod paymentMethod;

    @Column(name = "total_price", precision = 12, scale = 2)
    private BigDecimal totalPrice;

    @Column(name = "subtotal_amount", precision = 12, scale = 2)
    private BigDecimal subtotalAmount;

    @Column(name = "shipping_cost", precision = 12, scale = 2)
    private BigDecimal shippingCost;

    @Column(name = "discount_amount", precision = 12, scale = 2)
    private BigDecimal discountAmount;

    @Enumerated(EnumType.STRING)
    @Column(name = "order_status", columnDefinition = "enum('pending','processing','shipped','completed','cancelled') DEFAULT 'pending'")
    @Builder.Default
    private OrderStatus orderStatus = OrderStatus.pending;

    @Column(name = "order_time")
    @Builder.Default
    private LocalDateTime orderTime = LocalDateTime.now();

    @Column(name = "estimated_arrival")
    private LocalDateTime estimatedArrival;

    @Column(name = "receiver_name", length = 100)
    private String receiverName;

    @Column(name = "receiver_phone", length = 20)
    private String receiverPhone;

    @Column(name = "shipping_address", columnDefinition = "TEXT")
    private String shippingAddress;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<OrderDetail> orderDetails;

    public enum OrderStatus {
        pending, processing, shipped, completed, cancelled
    }
}
